includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -376,185 +376,233 @@ public function payment_fields() {
+					'payment_method'  => $this->get_payment_method_from_request(),
+					'token'           => $this->get_token_from_request(),
+					'is_saved_method' => empty( $_POST['wcpay-payment-method'] ) && ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-payment-token' ] ), // phpcs:ignore WordPress.Security.NonceVerification.Missing
+					'off_session'     => false,
+					'is_recurring'    => $is_recurring_payment,
+				]
+			);
+		} catch ( Exception $e ) {
+			// TODO: Create plugin specific exceptions so that we can be smarter about what we create notices for.
+			wc_add_notice( $e->getMessage(), 'error' );
+
+			$order->update_status( 'failed' );
+
+			return [
+				'result'   => 'fail',
+				'redirect' => '',
+			];
+		}
 	}
 
 	/**
 	 * Process the payment for a given order.
 	 *
-	 * @param WC_Order $order Order to process the payment for.
-	 * @param WC_Cart  $cart The WC_Cart object.
+	 * @param array $args Arguments that define this payment.
 	 *
 	 * @return array|null An array with result of payment and redirect URL, or nothing.
+	 * @throws WC_Payments_API_Exception Error processing the payment.
 	 */
-	public function process_payment_for_order( $order, $cart ) {
-		try {
-			$order_id = $order->get_id();
-			$amount   = $order->get_total();
-
-			if ( $amount > 0 ) {
-				// Get the payment method from the request (generated when the user entered their card details).
-				$payment_method      = $this->get_payment_method_from_request();
-				$manual_capture      = 'yes' === $this->get_option( 'manual_capture' );
-				$name                = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
-				$email               = sanitize_email( $order->get_billing_email() );
-				$save_payment_method = ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-new-payment-method' ] ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
-				$is_saved_method     = empty( $_POST['wcpay-payment-method'] ) && ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-payment-token' ] ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
-
-				// Determine the customer making the payment, create one if we don't have one already.
-				$user        = wp_get_current_user();
-				$customer_id = $this->customer_service->get_customer_id_by_user_id( $user->ID );
-
-				if ( null === $customer_id ) {
-					// Create a new customer.
-					$customer_id = $this->customer_service->create_customer_for_user( $user, $name, $email );
-				} else {
-					// Update the existing customer with the current details. In the event the old customer can't be
-					// found a new one is created, so we update the customer ID here as well.
-					$customer_id = $this->customer_service->update_customer_for_user( $customer_id, $user, $name, $email );
-				}
+	public function process_payment_for_order( $args ) {
+		// phpcs:disable Generic.Commenting.DocComment.MissingShort
```

Comment:
Can you think of a better way to document a "bag of options" argument? With these `@var` comments, at least my IDE picks it up.

Reply:
Maybe by encapsulating this in 1-3 classes? That might be a bit overkill at this point, but if we can reuse some of those classes somewhere else it might be worth it.

For now, the array is probably fine, and we can think about a refactor in a separate PR, _because_ we'd probably want to refactor more than just this portion of the code.

Reply:
cc @allendav @RadoslavGeorgiev @DanReyLop @luizreis This seems like it might be a bigger discussion surrounding subscriptions/orders as a whole. Passing in the big array definitely seems like a code smell.

Are there any classes we should be using to encapsulate this information? The array definitely feels a bit hacky, and there's probably some good way to encapsulate this information in 1 or more classes.

Off the top of my head, by looking at this for ~5 mins:

- `OrderInformation` -> `order` and `cart`.
- `PluginSettings` -> `manual_capture`.
- `PaymentMeta` -> `payment_method`, `token`, `is_saved_method`, `off_session`, `is_recurring`.

I feel like this might be a bigger discussion since this might add some complexity in the form of additional layers of abstraction. I'd welcome any thoughts on this.

Reply:
I agree that the array approach makes it easy to make mistakes - forgetting to pass an arg, spelling the arg key wrong, etc. I would be in favor of at least adding explicit shape tests to the top of process_payment_for_order but having classes would be even better for anything with more than one value (I wouldn't do it just for manual_capture)

Reply:
I'll have a quick look through the code then to see if there are any common patterns we can take advantage of here, and then create the classes to hold this information :) 

Since this might end up being a re-architecture of sorts I'm gonna copy the rest of @Automattic/hydra here 🙂

Reply:
I've done some initial work on this in 32ce51eaa7df7c8ad62cb49c511b777f4e3fcd28. I only extracted `payment_method`, `token`, `is_saved_method`, `off_session`, and `is_recurring` into a class I decided to call `WC_Payment_Information`. I did not introduce other classes in an effort not over-engineer this for now.

I want to do a little more work on this, specifically I'd like to get rid of the boolean parameters in the constructor and replace that with some constants.

```php
new WC_Payment_Information( $token->get_token(), $token, true, true, true );
```

looks horrible when compared to something like

```php
new WC_Payment_Information( $token->get_token(), $token, SAVED_CARD, RECURRING_PAYMENT, MERCHANT_INITIATED );
```

[SplEnum](https://www.php.net/manual/en/class.splenum.php) seems to be a good fit for this purpose? Or maybe it would be better to just create wrapper classes around some constants?

Reply:
I like the direction @reykjalin.

The array is flexible when it comes to adding more arguments in the future,  but I think your approach to using a single class allows us the same flexibility.

The wrapper class is very nicely named though I think we should namespace it and remove the `WC` prefix.

`SplEnum` would be exorbitant but using readable constants reads beautifully. I do wonder about the most readable way to namespace these constants.

Reply:
I agree with Dwain on SplEnum at this time. How about we start with the wrapper classes and that leaves a door open for SplEnum later.

This might be my embedded/C background leaking through, but would it be an idea to use OR'ing instead of passing a variable number of arguments, e.g

`Payment_Info_Option:SAVED_CARD | PaymentInfoOption:RECURRING | PaymentInfoOption:MERCHANT_INITIATED`

Reply:
> would it be an idea to use OR'ing instead of passing a variable number of arguments

That's your embedded/C background leaking through 😛 

On a more serious note, that's a good idea, _but_ it's a bit limiting if we want to use it for **all** the options in one parameter. It could be useful if we have an option such as `payment_type` where we want to _combine_ options. E.g.

```php
Payment_Information_Option::Recurring | Payment_Information_Option::PreOrder
```

or

```php
Payment_Information_Option::Single | Payment_Information_Option::PreOrder | Payment_Information_Option::PointOfSale
```

However, this will be very unreadable in the metadata, showing up as integers instead of human readable strings. It also feels a bit over-engineered—at least at the moment 🙂.

I'd suggest separate parameters (even if we were to go with OR-ing) and string constants for each. We then have the option to drop those parameters into the metadata, and someone reading the metadata will actually understand what the values mean.

Reply:
>I do wonder about the most readable way to namespace these constants.

I was thinking separate classes inside the same file as `Payment_Information`? Something like

```php
class Payment_Type {
    const SINGLE = 'single';
    const RECURRING = 'recurring';

    // We can add more later.
}
```

and

```php
class Payment_Method_Type {
    const SAVED_CARD = 'saved_card';
    const NEW_CARD = 'new_card';

    // Maybe we could later add more like 'check', 'cash', etc.
}
```

Names of classes are definitely up for debate, I'd love some feedback on those 🙂

Reply:
Namespace created in d33af671cc0c86c13e43832c5304297e82bee16e. I called it `WCPay\DataTypes`, but I'm open to a different name. It was honestly quite difficult to even come up with that one 😛

Reply:
I struggle with naming quite a bit @reykjalin!

I have one more suggestion :P 

What about leaving the payment information in the global `WCPay` namespace and then move the constants into `WCPay\Constants`?

This way we can always move collection classes like payment information into a different collection/namespace as we create more of them?

Reply:
Sounds good! Do you have any objections to wrapping the constants in relevant classes? I think a generic `Constants::SAVED_CARD` doesn't really give enough context to what the value relates to.

Reply:
> However, this will be very unreadable in the metadata, showing up as integers instead of human readable strings

Good point! Let's stick with strings. We probably will be looking at these in the Stripe dashboard frequently and doing hex math won't be fun for anyone (except me)

Reply:
>What about leaving the payment information in the global `WCPay` namespace

I've created a new issue for this: #864 

I think adding this would be a good fit for #860, so I'll mark this discussion as resolved and we can work on it after this PR and #832 have been merged 🙂